libsolidity/analysis/TypeChecker.cpp

```diff
@@ -2303,11 +2303,19 @@ bool TypeChecker::visit(FunctionCall const& _functionCall)
 		functionType = dynamic_cast<FunctionType const*>(expressionType);
 		funcCallAnno.kind = FunctionCallKind::FunctionCall;
 
+		if (auto memberAccess = dynamic_cast<MemberAccess const*>(&_functionCall.expression()))
+		{
+			if (dynamic_cast<FunctionDefinition const*>(memberAccess->annotation().referencedDeclaration))
+				_functionCall.expression().annotation().calledDirectly = true;
+		}
+		else if (auto identifier = dynamic_cast<Identifier const*>(&_functionCall.expression()))
+			if (dynamic_cast<FunctionDefinition const*>(identifier->annotation().referencedDeclaration))
+				_functionCall.expression().annotation().calledDirectly = true;
```

Comment:
Does this handle the `(f)()` case properly? I don't see anything to work around the fact that a single-element tuple inherits the type/declaration of the thing it contains.

Also, please add the test for `(L.foo)()`.

Reply:
It handles it by only handling `Identifier` and `MemberAccess` expressions so a `TupleExpression` is completely ignored. I basically copied this logic from the code gen.

Reply:
Makes sense.